src/colmap/scene/point2d.h

```diff
@@ -42,6 +42,9 @@ struct Point2D {
   // The image coordinates in pixels, starting at upper left corner with 0.
   Eigen::Vector2d xy = Eigen::Vector2d::Zero();
 
+  // The 2D covariance of the keypoint
+  Eigen::Matrix2d covar = Eigen::Matrix2d::Identity();
```

Comment:
What's the practical use case where we have a full 2D covariance? AFAIK, none of the existing keypoint detectors provides a meaningful 2D covariance? If we find a convincing use case, I suggest to parameterize this minimally with 3 floats and probably single-precision will be enough.

Reply:
Agree. There are different levels of detail possible:
1. identical for all points in an image (this would account for image resizing and different types of detectors)
2. isotropic (single scalar), 
3. anisotropic (2x2 matrix)

To me, it's not obvious that we need 3) directly. 2) has less impact on performance and is less invasive (using the `IsotropicNoiseCostFunctionWrapper`).

If 3), how would we obtain such covariances? In the case of SIFT, can we extract them from the affine shape? are they absolute in scale? For other detectors, I know about https://arxiv.org/pdf/2305.12250 but they are up to a scale factor.

Reply:
Yes. There are no current off-the-shelf detector that supports full 2D covariance. However, the current COLMAP does not support diagonal anisotropic covariance (case 1) from @sarlinpe) which is quite common from image resizing. Also, I have an internal development of 2D uncertainty in hand so would benefit from this update, and this will enable further development on such feature.

Reply:
Also, I have switched to 3-dim Vector3f in the Point2D class. PTAL : )

Reply:
For case 1, a single scalar per image would be sufficient, which has negligible storage overhead. For case 2, a single float per image point or, if using suitable quantization, a single uint8/uint16. For case 3, I would like to avoid supporting such experimental features in the main branch until they are proven to benefit the entire community. For classic blob/corner-based detectors (handcrafted and learned), I would expect an isotropic covariance to work quite well. They rarely ever detect features on elongated structures where the covariance is anisotropic? In this case, the viewing geometry of the supporting cameras to be the dominating factor for determining the covariance of 3D points?

Reply:
How about an image resized without ration kept? This could happen commonly in practice, which requires two parameters? I understand the point that full covariance might be too experimental to be supported, but this diagonal covariance would be quite practical (also generally falls into case 1))

Reply:
How about we revert the data flow but proceed with the support on the cost function only. This won't hurt anything but only give more flexible support : )